### PR TITLE
Elastic Processor : Process and Processes scraper 

### PR DESCRIPTION
**Pull request by ishleenk17**
**Description:** Added the processor and process scraper translation code in Elastic Processor

Closes : https://github.com/elastic/opentelemetry-dev/issues/168

### CODE REVIEW

File: cmd/otelcontribcol/go.mod [hunk 1]
```diff
@@ -2,7 +2,9 @@
 
 module github.com/open-telemetry/opentelemetry-collector-contrib/cmd/otelcontribcol
 
-go 1.20
+go 1.21.1
```
Comment (tommyers-elastic) [inline cmd/otelcontribcol/go.mod]: what's the reason for these changes?

File: connector/elasticconnector/connector.go [hunk 2]
```diff
@@ -4,8 +4,8 @@ import (
 	"context"
 	"strings"
 
-	"github.com/tommyers-elastic/opentelemetry-collector-contrib/connector/elasticconnector/internal/datastream"
-	"github.com/tommyers-elastic/opentelemetry-collector-contrib/connector/elasticconnector/internal/hostmetrics"
+	"github.com/ishleenk17/opentelemetry-collector-contrib/connector/elasticconnector/internal/datastream"
```
Comment (tommyers-elastic) [inline connector/elasticconnector/connector.go]: please revert these import changes

File: processor/elasticprocessor/internal/hostmetrics/hostmetrics.go [hunk 3]
```diff
@@ -38,6 +38,10 @@ func AddElasticSystemMetrics(scopeMetrics pmetric.ScopeMetrics) error {
 		return addMemoryMetrics(scopeMetrics.Metrics(), dataset)
 	case "load":
 		return addLoadMetrics(scopeMetrics.Metrics(), dataset)
+	case "process":
+		return addProcessMetrics(scopeMetrics.Metrics(), dataset)
+	case "processes":
+		return addProcessesMetrics(scopeMetrics.Metrics(), dataset)
```
Comment (tommyers-elastic) [inline processor/elasticprocessor/internal/hostmetrics/hostmetrics.go]: nit: should this function be called `addProcessSummaryMetrics`?

File: processor/elasticprocessor/internal/hostmetrics/process.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+	for i := 0; i < metrics.Len(); i++ {
+		metric := metrics.At(i)
+		if metric.Name() == "process.threads" {
+			dp := metric.Sum().DataPoints().At(0)
+			timestamp = dp.Timestamp()
+			threads = dp.IntValue()
+		} else if metric.Name() == "process.memory.utilization" {
+			dp := metric.Gauge().DataPoints().At(0)
+			timestamp = dp.Timestamp()
```
Comment (tommyers-elastic) [inline processor/elasticprocessor/internal/hostmetrics/process.go]: we are overriding timestamp for each case here. we can just set in the first case and assume all the datapoints have the same timestamp.

File: processor/elasticprocessor/internal/hostmetrics/process.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+			memVirtual = dp.IntValue()
+		} else if metric.Name() == "process.open_file_descriptors" {
+			dp := metric.Sum().DataPoints().At(0)
+			timestamp = dp.Timestamp()
+			fdOpen = dp.IntValue()
+		} else if metric.Name() == "process.cpu.time" {
+			dataPoints := metric.Sum().DataPoints()
+			for j := 0; j < dataPoints.Len(); j++ {
+				dp := dataPoints.At(j)
```
Comment (tommyers-elastic) [inline processor/elasticprocessor/internal/hostmetrics/process.go:34]: nit: the formatting looks off here, please run `go fmt`

Comment (tommyers-elastic) [reply processor/elasticprocessor/internal/hostmetrics/process.go:34]: maybe it's just way it's being displayed in this diff. if i'm wrong, apologies in advance.

File: processor/elasticprocessor/internal/hostmetrics/process.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+				timestamp = dp.Timestamp()
+				value := dp.DoubleValue()
+				if state, ok := dp.Attributes().Get("state"); ok {
```
Comment (tommyers-elastic) [inline processor/elasticprocessor/internal/hostmetrics/process.go:37]: is there any need for this switch statement. all the cases are the same, i think we can replace it all with just `total += dp.DoubleValue()`

File: processor/elasticprocessor/internal/hostmetrics/process.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+						total += value
+
+					}
+				}
+			}
+		}
+	}
+
+	memUtilPct := memUtil / 100
+	cpuTimeValue = total * 1000
```
Comment (tommyers-elastic) [inline processor/elasticprocessor/internal/hostmetrics/process.go:83]: tiny nit: why not define both of these values here with `:=` vs one at the top and one here

File: processor/elasticprocessor/internal/hostmetrics/process.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+			dataType:  Sum,
+			name:      "system.process.fd.open",
+			timestamp: timestamp,
+			intValue:  &fdOpen,
+		},
+		metric{
+			dataType:    Gauge,
+			name:        "process.memory.pct",
+			timestamp:   timestamp,
+			doubleValue: &memUtilPct,
```
Comment (tommyers-elastic) [inline processor/elasticprocessor/internal/hostmetrics/process.go:123]: is it correct that this is always the same value as `system.process.memory.rss.pct`?

File: processor/elasticprocessor/internal/hostmetrics/process.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+		},
+		metric{
+			dataType:  Sum,
+			name:      "system.process.fd.open",
+			timestamp: timestamp,
+			intValue:  &fdOpen,
+		},
+		metric{
+			dataType:    Gauge,
+			name:        "process.memory.pct",
```
Comment (tommyers-elastic) [inline processor/elasticprocessor/internal/hostmetrics/process.go:121]: missing `system.` prefix in the metric name

File: processor/elasticprocessor/internal/hostmetrics/processes.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+					case "zombies":
+						zombieProcesses = value
+						total += value
+					}
+				}
+			}
+		}
+
+	}
+	totalProcesses = total
```
Comment (tommyers-elastic) [inline processor/elasticprocessor/internal/hostmetrics/processes.go]: this can be removed, let's just use `totalProcesses` throughout, it's more descriptive

File: processor/elasticprocessor/internal/hostmetrics/processes.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
+package hostmetrics
+
+import (
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+func addProcessesMetrics(metrics pmetric.MetricSlice, dataset string) error {
+	var timestamp pcommon.Timestamp
+	var total, value, idleProcesses, sleepingProcesses, stoppedProcesses, zombieProcesses, totalProcesses int64
```
Comment (tommyers-elastic) [inline processor/elasticprocessor/internal/hostmetrics/processes.go]: why is `value` defined here?

**Review by tommyers-elastic — COMMENTED** [review]
couple of small things before it's ready to merge - thanks!

File: processor/elasticprocessor/internal/hostmetrics/processes.go (hunk 10 above)
Comment (ishleenk17) [reply processor/elasticprocessor/internal/hostmetrics/processes.go]: Done

File: processor/elasticprocessor/internal/hostmetrics/processes.go (hunk 11 above)
Comment (ishleenk17) [reply processor/elasticprocessor/internal/hostmetrics/processes.go]: Taken care

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 9 above)
Comment (ishleenk17) [reply processor/elasticprocessor/internal/hostmetrics/process.go:121]: The metric name is process.memory.pct

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 7 above)
Comment (ishleenk17) [reply processor/elasticprocessor/internal/hostmetrics/process.go:83]: Taken care

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 6 above)
Comment (ishleenk17) [reply processor/elasticprocessor/internal/hostmetrics/process.go:37]: I had to add new metrics under each of the case, which you can see in the latest diff. 
Hence, the swicth statement

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 5 above)
Comment (ishleenk17) [reply processor/elasticprocessor/internal/hostmetrics/process.go:34]: Yeah, it's properly formatted. Ran go fmt also. But no change.
I think its the way diff shows up

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 4 above)
Comment (ishleenk17) [reply processor/elasticprocessor/internal/hostmetrics/process.go]: What if it doesn't enter the first if loop statement?

File: processor/elasticprocessor/internal/hostmetrics/hostmetrics.go (hunk 3 above)
Comment (ishleenk17) [reply processor/elasticprocessor/internal/hostmetrics/hostmetrics.go]: Yes, this name is better

File: connector/elasticconnector/connector.go (hunk 2 above)
Comment (ishleenk17) [reply connector/elasticconnector/connector.go]: I was facing issues while running the changes. Hence I had to change it. 
Actually, shall we remove the connector code from this PR ?
Its not relevant

File: cmd/otelcontribcol/go.mod (hunk 1 above)
Comment (ishleenk17) [reply cmd/otelcontribcol/go.mod]: Hmmm, I ran a go mod tidy and go get. 
Not sure if that updates these?

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 8 above)
Comment (ishleenk17) [reply processor/elasticprocessor/internal/hostmetrics/process.go:123]: I have checked across multiple documents for different processes. 
I have noticed the values to be always the same.

We can verify from the logic written in beats for it once

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 4 above)
Comment (tommyers-elastic) [reply processor/elasticprocessor/internal/hostmetrics/process.go]: there are lots of assumptions in the processor about the format of the input data. we know that these metrics are contained in the same `ScopeMetrics`. As long as we pull the timestamp from a default metric, not one that has to be enabled in config, i think we are fine.

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 5 above)
Comment (tommyers-elastic) [reply processor/elasticprocessor/internal/hostmetrics/process.go:34]: ok thanks! sorry!

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 8 above)
Comment (tommyers-elastic) [reply processor/elasticprocessor/internal/hostmetrics/process.go:123]: weird! perhaps it's worth a comment here, it almost looks like a mistake.

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 9 above)
Comment (tommyers-elastic) [reply processor/elasticprocessor/internal/hostmetrics/process.go:121]: aha 👍

File: connector/elasticconnector/connector.go (hunk 2 above)
Comment (tommyers-elastic) [reply connector/elasticconnector/connector.go]: you can get around the naming issue by using a `replace`, there is an example in the config already.

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 4 above)
Comment (ishleenk17) [reply processor/elasticprocessor/internal/hostmetrics/process.go]: @tommyers-elastic : Actually,  there is a possibility that a user can disable a particular default metric as well by doing this. 

```
metrics:
  <metric_name>:
    enabled: false
```

In that case, if the metric where we specify timestamp is disabled, we will not get timestamp for other metrics as well, I suppose

File: cmd/otelcontribcol/go.mod (hunk 1 above)
Comment (ishleenk17) [reply cmd/otelcontribcol/go.mod]: Updated the paths

File: connector/elasticconnector/connector.go (hunk 2 above)
Comment (ishleenk17) [reply connector/elasticconnector/connector.go]: Ok, I'll try it this way

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 4 above)
Comment (tommyers-elastic) [reply processor/elasticprocessor/internal/hostmetrics/process.go]: yeh i think we can assume that if the customer does not apply the configurations we tell them the processor requires, they can expect it not to work

that said, timestamp is super important. let's merge this as is, and i will put write some common code that all these functions can use that sets the timestamp if not set.

**Review by tommyers-elastic — APPROVED** [review]

File: processor/elasticprocessor/internal/hostmetrics/process.go (hunk 8 above)
Comment (ishleenk17) [reply processor/elasticprocessor/internal/hostmetrics/process.go:123]: @tommyers-elastic 
Checked the logic in beats. 
See [here](https://github.com/elastic/elastic-agent-system-metrics/blob/main/metric/system/process/process_types.go#L164). It is indeed equal to rs.pct
